Support multi-tenant use of the gem

petergoldstein opened the pull request:
Addresses #149 

This is a quick and dirty conversion to multi-tenancy for the gem.  That said, both Rubocop and specs pass and I expect it works in practice.

Some changes I'd suggest on top of this commit:

1. Make the Files, Finetunes, Images, and Models classes non-instantiable except via the Client object.  All access to these APIs should go through a Client object.  This would dramatically simplify the initializers for these classes
2. Potentially add a `default_client` class method to OpenAI::Client, which would produce a memoized OpenAI::Client object that is initialized with the global parameters
3. Consider swapping out Httparty for connection pools, to minimize latency under production conditions
4. Maybe exclude the organization header from the header set if the value is nil?  I don't think we need to send in that case.
5. Allow the global configuration to have a missing access key, and shift that error to the Client upon initialization.

I don't think these changes make the gem any more difficult to use.  The only change from a user perspective is that setting the API credentials on a single client doesn't change it for the whole app.  So if you don't set the global credentials, you will need to invoke every Client instance with credentials.

## All Submissions:

* [X] Have you followed the guidelines in our [Contributing document](../blob/main/CONTRIBUTING.md)?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?
* [X] Have you added an explanation of what your changes do and why you'd like us to include them?

alexrudall commented:
@petergoldstein thanks a lot for this! It looks great.

If I run some of the specs without VCR (so actually hitting the API, this is a new feature I added last week) 

```
  NO_VCR=true bundle exec rspec spec/ruby/openai/client/completions/codex_spec.rb
```

I get failures with this error

```
 "message": "No such organization: <OPENAI_ACCESS_TOKEN>.",
```

Do you know why this is?

petergoldstein commented:
@alexrudall Yes, there was a bug in the code (I had fixed it in the other PR but not here) where the wrong default value was being used when the organization_id is not set.  I just updated it here.  There are specs in the other PR that catch this, but since I moved the code around a bit it's not a trivial pull-back.

bf4 commented:
@petergoldstein @alexrudall looks good, though needs a rebase against main now.  Any blockers to this?

bf4 commented:
@alexrudall any way to revive this PR? The Twitter gem did the same thing many years ago-- switched from being a global client to an instantiatable client.  I think it's probably a good pattern and already has a contribution.

bf4 commented:
I'm happy to revive this if there's still interest. I think it's good work.

bf4 commented:
This has been open since January with maintainer interest. I recommend closing.

MarioRuiz commented:
right now we don't have any option to have client instances with different hosts and/or tokens so I think this is really useful and necessary if we want the gem to be used widely.

petergoldstein commented:
@MarioRuiz If you need another option for handling multitenant OpenAI access you can check out the gem I created.  I don't want to abuse another repos comments, so I won't link directly, but it's easy enough to find.

alexrudall commented:
OK, let's do this - thanks for patience everyone. Working with it on a separate branch where I've done the rebase: #260

alexrudall commented:
Closing for #260
